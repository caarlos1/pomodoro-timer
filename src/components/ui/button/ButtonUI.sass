.button-ui
    background: #605BFF
    padding: 10px 20px
    border: none
    color: #FFF
    border-radius: 8px
    transition: 0.1s
    height: 40px

    &:hover
        background: #554fff
        cursor: pointer
        filter: drop-shadow(4px 4px 0px #1e1a6134)

    &:active
        transform: translateY(2px)

    &.full
        width: 100%

    &:disabled
        opacity: .5
        filter: none

        &:active
            transform: none

        &:hover
            background: #605BFF
            cursor: default
