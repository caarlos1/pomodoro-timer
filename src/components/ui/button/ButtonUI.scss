.button-ui {
  background: var(--action-secondary);
  color: var(--text-color);
  padding: 14px 20px;
  border: none;
  transition: all 0.1s ease-in-out;
  font-size: 1.6rem;
  line-height: 1;
  border-radius: 5px;

  &:hover {
    cursor: pointer;
    background: var(--action-secondary-dark);
  }

  &:active {
    transform: translateY(2px);
  }

  &:disabled {
    opacity: 0.5;
    filter: none;

    &:hover {
      cursor: default;
      background: var(--action-secondary);
    }

    &:active {
      transform: none;
    }
  }

  &.sharp {
    border-radius: 0;
  }

  &.full {
    width: 100%;
  }

  &.lg {
    padding: 20px 25px;
    font-size: 1.7rem;
  }

  &.sm {
    padding: 10px 15px;
    font-size: 1.5rem;
  }

  &.primary {
    color: #fff;
    background: var(--action-primary);

    &:hover {
      background: var(--action-primary-dark);

      &:disabled {
        background: var(--action-primary);
      }
    }
  }

  &.confirm {
    color: #fff;
    background: var(--action-confirm);

    &:hover {
      background: var(--action-confirm-dark);

      &:disabled {
        background: var(--action-confirm);
      }
    }
  }

  &.warning {
    color: #fff;
    background: var(--action-warning);

    &:hover {
      background: var(--action-warning-dark);

      &:disabled {
        background: var(--action-warning);
      }
    }
  }

  &.remove {
    color: #fff;
    background: var(--action-remove);

    &:hover {
      background: var(--action-remove-dark);

      &:disabled {
        background: var(--action-remove);
      }
    }
  }

  &.help {
    color: #fff;
    background: var(--action-help);

    &:hover {
      background: var(--action-help-dark);

      &:disabled {
        background: var(--action-help);
      }
    }
  }
}
