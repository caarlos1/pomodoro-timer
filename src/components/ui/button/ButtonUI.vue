<script setup lang="ts">
import { computed, useSlots } from "vue";

export type ButtonUIProps = {
  label?: string;
  design?: string;
  full?: boolean;
  sharp?: boolean;
  disabled?: boolean;
  size?: string;
};

const props = withDefaults(defineProps<ButtonUIProps>(), {
  label: "",
  design: "",
  full: false,
  disabled: false,
  sharp: false,
  size: "sm",
});

const slots = useSlots();

const computedClasses = computed(() => {
  const { full, size, sharp, disabled, design } = props;
  return {
    full,
    [size]: size.length,
    sharp,
    disabled,
    [design]: design.length,
  };
});
</script>

<template>
  <button class="button-ui" :class="computedClasses" :disabled="disabled">
    <template v-if="slots['default']">
      <slot></slot>
    </template>
    <template v-else>
      {{ label }}
    </template>
  </button>
</template>

<style lang="scss" scoped>
@import "./ButtonUI.scss";
</style>
