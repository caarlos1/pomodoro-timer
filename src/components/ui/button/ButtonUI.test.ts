import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";

import ButtonUI from ".";

describe("ButtonUI", () => {
  it("should render a button with correct label", () => {
    const wrapper = mount(ButtonUI, { props: { label: "Simple Text" } });
    expect(wrapper.text()).toContain("Simple Text");
  });

  it("should render a button with full style", () => {
    const wrapper = mount(ButtonUI, { props: { full: true } });
    expect(wrapper.classes()).toContain("full");
  });

  it("should render a slot item", () => {
    const wrapper = mount(ButtonUI, {
      slots: { default: "<p>Slot Example</p>" },
    });

    expect(wrapper.text()).toContain("Slot Example");
  });

  it("should match a snapshot", () => {
    const { element } = mount(ButtonUI, {
      props: {
        label: "Snap! ;)",
        design: "some simple classes",
        sharp: true,
        full: true,
        disabled: true,
      },
    });
    expect(element).toMatchSnapshot();
  });
});
