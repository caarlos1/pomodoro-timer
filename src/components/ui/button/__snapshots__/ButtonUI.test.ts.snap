// Vitest Snapshot v1

exports[`ButtonUI > should match a snapshot 1`] = `
<button
  class="button-ui full sm sharp disabled some simple classes"
  data-v-a4d02f90=""
  disabled="true"
>
  
  Snap! ;)
  
</button>
`;
